### PR TITLE
Added a hat drop check.

### PR DESCRIPTION
**Pull request by ccuser44**
This stops exploiters from dropping hats. Some games have hat drop scripts so I added a setting called CanHatsBeDropped. It is defaulted to false as most games do not have a hat drop script but if they do they can set it to true.
If you find any bugs or known a way to do this shorter plz tell me. ;)

### CODE REVIEW

**Review by Hexcede — CHANGES_REQUESTED** [review]
Rather than `:Destroying()` hats I think it makes more sense to parent it back to the character. Plus, you also have the check for `:Destroy()` being used by the client only to destroy the hat anyway. I'm assuming this was probably a typo.

Additionally, rather than the name `CanHatsBeDropped` I would prefer something like `InvalidAccessoryMovement`, as, the check (in the way you committed it) encompasses both dropping and deletion, which I think is valid, unlike with tools where there may be more compatibility issues due to scripts.

**Comment by ccuser44** [comment]
Ok. Changed the setting name for hat dropped to CanParentAccroutrements as it applies both to hats and accessories. Also added a setting where you can toggle if invalidly dropped accroutrements get deleted. By deafault this is on to prevent spam parenting them to workspace which could cause lag and to prevent hacky stuff.

**Comment by ccuser44** [comment]
Ok fixed some bugs which my code has. Also named it PreventParentAccroutrements so players are not confused about the naming like the previous one was confusing because it was false and they could think that turning it to true could turn the check on instead of off. Ready to merge but if you would like me to change something i'd be happy to do that :D

File: Anticheat/src/init.lua [hunk 1]
```diff
@@ -60,6 +61,8 @@ Anticheat.ChecksEnabled = {
 	InvalidDrop = true, -- Dropping tools that don't have CanBeDropped
 	ToolDeletion = true, -- Stop the client from deleting tools (Incompatible with any usage of tool.Parent = nil, use :Destroy() instead)
 	FEGodMode = true, -- God mod achieved by deleting their Humanoid on the server and creating a fake one on the client
+	PreventParentAccroutrements = true, -- If players can drop and/or delete hats(and other accessories). If you have hat drop scripts turn this to false. Up to 2017 you could drop hats via the = key however this was removed. If you have a custom hat drop script set this to false.
```
Comment (ghost) [inline Anticheat/src/init.lua]: says if you have hat scripts turn this off 2 times

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
also, you can remove shit inside your hat's server-sided from client(%100 tested) every hat on roblox has a mesh inside it and if you remove it your hat becomes a literal block which i dont know if you check for it but if you dont then this pull request is useless since people who drop hats on workspace also remove meshes inside them

tldr: check if they remove instance's inside our hat

**Comment by Hexcede** [comment]
> also, you can remove shit inside your hat's server-sided from client(%100 tested) every hat on roblox has a mesh inside it and if you remove it your hat becomes a literal block which i dont know if you check for it but if you dont then this pull request is useless since people who drop hats on workspace also remove meshes inside them
> 
> tldr: check if they remove instance's inside our hat

I don't see how not checking for the mesh being deleted makes this check useless. The check is not bypassed simply because an exploiter deleted the mesh. The point of this is to prevent exploiters from creating a giant mess and using up server memory by dropping loads of hats.

Any descendant of the character can be deleted by the client and it will replicate, there are (for real) no exceptions to this rule whatsoever, as long as the instance is a descendant of `Player.Character` and the two models match on the client and server, deletion events replicate 100% of the time. Including descendants of equipped tools. Unfortunately there aren't good ways to check for these sorts of things.

As for deleting the mesh from the hat, this is purely a visual change. It doesn't change collisions or any other behavior, I would consider it a minimal issue. This is sort of like the client deleting the `Handle` from a tool. While yes, it can cause undesirable effects, it isn't particularly abusable under normal circumstances (deleting the mesh is probably one of the least harmful things you can do with that replication quirk especially)

Also, just thought I'd mention, currently there is an extra r in Accoutrements: Acc**r**outrements

**Comment by Hexcede** [comment]
I think this pull request differs too much from how I would prefer to implement this feature, so, I'm not going to merge it, however, I will (next time I publish an update to the anticheat) create my own implementation of this check.

**Comment by ccuser44** [comment]
Ok. Removed extra r in accoutrements. Renamed some stuff. And removed the second time where it said set this to false if you have a custom hat drop script.

**Comment by ccuser44** [comment]
One thing I am wondering if you are going to reimplement this will you add the :Destroying thing there as a setting or not?

**Comment by ccuser44** [comment]
Anyways if you don't want to completely rewrite it you can edit my patch (patch-2) as I enabled editing my maintainers. And then merge it.

File: Anticheat/src/init.lua (hunk 1 above)
Comment (ccuser44) [reply Anticheat/src/init.lua]: Ok fixed it.

**Comment by ccuser44** [comment]
Ok maybe the deletion of hats by default wasn't such a good idea. I'll turn the default off. If you think I should remove it totally please let me known. ;)

**Comment by ccuser44** [comment]
@Hexcede I removed the deletion code totally as I decided it was unnecessary.

**Comment by ccuser44** [comment]
Ok so I totally remade how this works. So it works like the tools do too. There is PreventAccoutrementDrop and AccoutrementDeletion. They works seperately of each other. The first prevents the dropping of accoutrements and the second one prevents deletion. So it basically works how you should expect it to. Also I removed the weird :Destroy() thing there.

**Review by Hexcede — APPROVED** [review]

**Comment by Hexcede** [comment]
The projects under the main repo have been turned into sub modules rather than sub folders. The Anticheat repo is here: https://github.com/Hexcede/Hexolus-Anticheat

**Comment by ccuser44** [comment]
Ok remade this pullrequest to https://github.com/Hexcede/Hexolus-Anticheat/pull/6
